### PR TITLE
Expose underlying unix file descriptor in SafeMemoryMappedFileHandle

### PR DESCRIPTION
**Pull request by dannyfriar**
See https://github.com/dotnet/runtime/issues/52854

Done in a similar way to [SafePipeHandle](https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.IO.Pipes/src/Microsoft/Win32/SafeHandles/SafePipeHandle.Unix.cs#L106).

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=53538) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=53538) before we can accept your contribution.<br/><br/>:x: dannyfriar [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=53538)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=53538) it.</sub>

File: src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs [hunk 1]
```diff
@@ -59,9 +57,22 @@ public sealed partial class SafeMemoryMappedFileHandle : SafeHandleZeroOrMinusOn
             _options = options;
             _capacity = capacity;
 
-            // Fake a unique int handle value > 0.
-            int nextHandleValue = (int)((Interlocked.Increment(ref s_counter) % (int.MaxValue - 1)) + 1);
-            SetHandle(new IntPtr(nextHandleValue));
+            IntPtr handlePtr;
+
+            if (fileStream != null) {
+                _fileStreamHandle = fileStream.SafeFileHandle;
```
Comment (stephentoub) [inline src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:63]: This should really be stored only after calling DangerousAddRef.  The issue is that if DangerousAddRef throws, the finalizer for this object will still be invoked, and if it sees _fileStreamHandle as non-null, it'll call DangerousRelease.  Now, I see we do this same ordering in pipes, and in reality it's unlikely to be an actual issue, since the only reason DangerousAddRef should throw is if the handle has already been disposed, in which case an extraneous DangerousRelease on it in the finalizer shouldn't matter functionally.  But, still good hygiene.

File: src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs [hunk 2]
```diff
@@ -59,9 +57,22 @@ public sealed partial class SafeMemoryMappedFileHandle : SafeHandleZeroOrMinusOn
...
+
+                bool ignored = false;
+                _fileStreamHandle.DangerousAddRef(ref ignored);
+
+                handlePtr = _fileStreamHandle.DangerousGetHandle();
+            }
+            else {
+                _fileStreamHandle = null;
```
Comment (stephentoub) [inline src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:71]: Why is this needed?

File: src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs [hunk 3]
```diff
@@ -74,7 +85,17 @@ protected override void Dispose(bool disposing)
             base.Dispose(disposing);
         }
 
-        protected override bool ReleaseHandle() => true; // Nothing to clean up.  We unlinked immediately after creating the backing store.
+        protected override bool ReleaseHandle() {
+            Debug.Assert(!IsInvalid);
```
Comment (stephentoub) [inline src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:89]: This assert isn't valid.  The ctor could throw before reaching the SetHandle call, in which case the handle will still have a value of 0, which is considered invalid.

File: src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs [hunk 4]
```diff
@@ -74,7 +85,17 @@ protected override void Dispose(bool disposing)
...
+
+            if (_fileStreamHandle != null) {
+                SetHandle((IntPtr) (-1));
+                _fileStreamHandle.DangerousRelease();
+                _fileStreamHandle = null;
+            }
+
+            return true;
+        }
```
Comment (stephentoub) [inline src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:98]: Can you please add test validating the desired functionality?

Comment (dannyfriar) [reply src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:95]: Sure, I will add

Comment (dannyfriar) [reply src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:95]: @stephentoub I've added a test to check it matches the filestream handle

File: src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs [hunk 5]
```diff
@@ -59,9 +57,21 @@ public sealed partial class SafeMemoryMappedFileHandle : SafeHandleZeroOrMinusOn
...
+
+            if (fileStream != null)
+            {
+                bool ignored = false;
+                fileStream.SafeFileHandle.DangerousAddRef(ref ignored);
+
+                _fileStreamHandle = fileStream.SafeFileHandle;
+                handlePtr = _fileStreamHandle.DangerousGetHandle();
+            }
+            else {
```
Comment (stephentoub) [inline src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:70]: Nit:
```suggestion
            else
            {
```

File: src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs [hunk 6]
```diff
@@ -74,7 +84,15 @@ protected override void Dispose(bool disposing)
             base.Dispose(disposing);
         }
 
-        protected override bool ReleaseHandle() => true; // Nothing to clean up.  We unlinked immediately after creating the backing store.
+        protected override bool ReleaseHandle() {
+            if (_fileStreamHandle != null) {
```
Comment (stephentoub) [inline src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:88]:
```suggestion
            if (_fileStreamHandle != null)
            {
```

File: src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs [hunk 7]
```diff
@@ -74,7 +84,15 @@ protected override void Dispose(bool disposing)
             base.Dispose(disposing);
         }
 
-        protected override bool ReleaseHandle() => true; // Nothing to clean up.  We unlinked immediately after creating the backing store.
+        protected override bool ReleaseHandle() {
```
Comment (stephentoub) [inline src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:87]:
```suggestion
        protected override bool ReleaseHandle()
        {
```

File: src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs [hunk 8]
```diff
@@ -59,9 +57,21 @@ public sealed partial class SafeMemoryMappedFileHandle : SafeHandleZeroOrMinusOn
...
-            SetHandle(new IntPtr(nextHandleValue));
+            IntPtr handlePtr;
+
+            if (fileStream != null)
+            {
+                bool ignored = false;
+                fileStream.SafeFileHandle.DangerousAddRef(ref ignored);
+
+                _fileStreamHandle = fileStream.SafeFileHandle;
+                handlePtr = _fileStreamHandle.DangerousGetHandle();
```
Comment (stephentoub) [inline src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:68]:
```suggestion
                SafeFileHandle handle = fileStream.SafeFileHandle;
                handle.DangerousAddRef(ref ignored);
                _fileStreamHandle = handle;
                handlePtr = handle.DangerousGetHandle();
```

File: src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs [hunk 9]
```diff
@@ -949,5 +949,21 @@ public void OpenCharacterDeviceAsFile(MemoryMappedFileAccess access)
...
+
+        /// <summary>
+        /// Test to verify that the MemoryMappedFile has the same underlying handle as the FileStream it's created from
+        /// </summary>
+        [PlatformSpecific(TestPlatforms.AnyUnix)]
+        [Fact]
+        public void MapHandleMatchesFileStreamHandle()
+        {
+            using (FileStream fs = File.OpenWrite(GetTestFilePath())) {
+                MemoryMappedFile mmf = MemoryMappedFile.CreateFromFile(
```
Comment (stephentoub) [inline src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs:961]: This should be disposed as well.

File: src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs [hunk 10]
```diff
@@ -11,15 +11,13 @@ namespace Microsoft.Win32.SafeHandles
...
-        private static long s_counter;
-
         /// <summary>
         /// The underlying FileStream.  May be null.  We hold onto the stream rather than just
         /// onto the underlying handle to ensure that logic associated with disposing the stream
         /// (e.g. deleting the file for DeleteOnClose) happens at the appropriate time.
         /// </summary>
-        internal readonly FileStream? _fileStream;
+        private readonly FileStream? _fileStream;
+        internal SafeFileHandle? _fileStreamHandle;
```
Comment (stephentoub) [inline src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:20]:
```suggestion
        /// <summary>The FileStream's handle, cached to avoid repeated accesses to FileStream.SafeFileHandle that could, in theory, change.</summary>
        internal SafeFileHandle? _fileStreamHandle;
```

**Review by stephentoub — APPROVED** [review]
Left some comments to be addressed, but otherwise looks good. Thanks.

**Comment by dannyfriar** [comment]
> Left some comments to be addressed, but otherwise looks good. Thanks.

Thanks @stephentoub. I've addressed your comments (the test failures are unrelated I think).

**Review by stephentoub — APPROVED** [review]
